File: backend/validate.py
from __future__ import annotations

from .models import Bucket, Issue, Project, RectMM


def _bucket_body_rect(b: Bucket) -> tuple[float, float, float, float]:
    return (b.body_mm.x, b.body_mm.y, b.body_mm.x + b.body_mm.w, b.body_mm.y + b.body_mm.d)


def _bucket_base_rect_mm(b: Bucket, cell_mm: float) -> tuple[float, float, float, float]:
    bx = b.base_cells.x * cell_mm
    by = b.base_cells.y * cell_mm
    return (bx, by, bx + b.base_cells.w * cell_mm, by + b.base_cells.d * cell_mm)


def _overlaps(a: tuple[float, float, float, float], b: tuple[float, float, float, float]) -> bool:
    return not (a[2] <= b[0] or b[2] <= a[0] or a[3] <= b[1] or b[3] <= a[1])


def validate_project(project: Project) -> list[Issue]:
    issues: list[Issue] = []
    cell = project.grid.cell_mm
    drawer_rect = (0.0, 0.0, project.drawer.width_mm, project.drawer.depth_mm)

    for i, b in enumerate(project.buckets):
        body = _bucket_body_rect(b)
        base = _bucket_base_rect_mm(b, cell)

        if not (drawer_rect[0] <= body[0] and body[2] <= drawer_rect[2]
                and drawer_rect[1] <= body[1] and body[3] <= drawer_rect[3]):
            issues.append(Issue(severity="warning", bucket_id=b.id,
                                code="BODY_OUT_OF_DRAWER",
                                message="Bucket body extends beyond drawer."))

        if not (drawer_rect[0] <= base[0] and base[2] <= drawer_rect[2]
                and drawer_rect[1] <= base[1] and base[3] <= drawer_rect[3]):
            issues.append(Issue(severity="error", bucket_id=b.id,
                                code="BASE_OUT_OF_DRAWER",
                                message="Gridfinity base extends beyond drawer."))

        exceeds_xy = b.body_mm.w > project.printer.bed_x_mm or b.body_mm.d > project.printer.bed_y_mm
        exceeds_z = b.height_mm > project.printer.bed_z_mm
        if exceeds_xy and b.split.enabled and b.split.strategy == "naive":
            issues.append(Issue(severity="info", bucket_id=b.id,
                                code="NAIVE_SPLIT_ENABLED",
                                message="Bucket will export as multiple naive split STLs."))
        elif exceeds_xy or exceeds_z:
            issues.append(Issue(severity="warning", bucket_id=b.id,
                                code="EXCEEDS_PRINTER_BED",
                                message="Bucket exceeds printer bed; needs split."))

        if exceeds_z and b.split.enabled:
            issues.append(Issue(severity="warning", bucket_id=b.id,
                                code="HEIGHT_EXCEEDS_PRINTER_BED",
                                message="Naive split only splits X/Y; bucket height still exceeds printer bed."))

        for j, other in enumerate(project.buckets):
            if j <= i:
                continue
            obody = _bucket_body_rect(other)
            obase = _bucket_base_rect_mm(other, cell)
            if _overlaps(body, obody):
                issues.append(Issue(severity="error", bucket_id=b.id,
                                    code="BODY_OVERLAP",
                                    message=f"Bucket body overlaps {other.id}."))
            if _overlaps(base, obase):
                issues.append(Issue(severity="error", bucket_id=b.id,
                                    code="BASE_OVERLAP",
                                    message=f"Gridfinity base overlaps {other.id}."))

    return issues

File: web/src/validate.js
// JS port of backend/validate.py — bounds, collision, and printer-bed checks.

function bodyRect(b) {
  return [b.body_mm.x, b.body_mm.y, b.body_mm.x + b.body_mm.w, b.body_mm.y + b.body_mm.d];
}
function baseRectMm(b, cellMm) {
  const x = b.base_cells.x * cellMm;
  const y = b.base_cells.y * cellMm;
  return [x, y, x + b.base_cells.w * cellMm, y + b.base_cells.d * cellMm];
}
function overlaps(a, b) {
  return !(a[2] <= b[0] || b[2] <= a[0] || a[3] <= b[1] || b[3] <= a[1]);
}
function inside(rect, drawer) {
  return drawer[0] <= rect[0] && rect[2] <= drawer[2] && drawer[1] <= rect[1] && rect[3] <= drawer[3];
}

export function validateProject(project) {
  const issues = [];
  const cell = project.grid.cell_mm;
  const drawer = [0, 0, project.drawer.width_mm, project.drawer.depth_mm];
  const printer = project.printer;

  for (let i = 0; i < project.buckets.length; i++) {
    const b = project.buckets[i];
    const body = bodyRect(b);
    const base = baseRectMm(b, cell);

    if (!inside(body, drawer)) {
      issues.push({
        severity: "warning", bucket_id: b.id,
        code: "BODY_OUT_OF_DRAWER",
        message: "Bucket body extends beyond drawer.",
      });
    }
    if (!inside(base, drawer)) {
      issues.push({
        severity: "error", bucket_id: b.id,
        code: "BASE_OUT_OF_DRAWER",
        message: "Gridfinity base extends beyond drawer.",
      });
    }

    const exceedsXy = b.body_mm.w > printer.bed_x_mm || b.body_mm.d > printer.bed_y_mm;
    const exceedsZ = b.height_mm > printer.bed_z_mm;
    const naive = b.split && b.split.enabled && b.split.strategy === "naive";
    if (exceedsXy && naive) {
      issues.push({
        severity: "warning", bucket_id: b.id,
        code: "NAIVE_SPLIT_ENABLED",
        message: "Bucket exceeds printer bed; will export as multiple naive split parts. Uncheck “Naive split” to disable.",
      });
    } else if (exceedsXy || exceedsZ) {
      issues.push({
        severity: "warning", bucket_id: b.id,
        code: "EXCEEDS_PRINTER_BED",
        message: "Bucket exceeds printer bed and split is disabled — export will fail.",
      });
    }
    if (exceedsZ && b.split && b.split.enabled) {
      issues.push({
        severity: "warning", bucket_id: b.id,
        code: "HEIGHT_EXCEEDS_PRINTER_BED",
        message: "Naive split only splits X/Y; bucket height still exceeds printer bed.",
      });
    }

    for (let j = i + 1; j < project.buckets.length; j++) {
      const other = project.buckets[j];
      if (overlaps(body, bodyRect(other))) {
        issues.push({
          severity: "error", bucket_id: b.id,
          code: "BODY_OVERLAP",
          message: `Bucket body overlaps ${other.id}.`,
        });
      }
      if (overlaps(base, baseRectMm(other, cell))) {
        issues.push({
          severity: "error", bucket_id: b.id,
          code: "BASE_OVERLAP",
          message: `Gridfinity base overlaps ${other.id}.`,
        });
      }
    }
  }
  return issues;
}

export function isValid(issues) {
  return !issues.some((i) => i.severity === "error");
}
